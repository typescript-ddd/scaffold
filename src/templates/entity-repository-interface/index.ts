export * from "./entity-repository-interface";