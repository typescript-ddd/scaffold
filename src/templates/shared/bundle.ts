import { GenerateContext, Chunk } from "./generate-context";
import { Template, TemplateOption } from "./template";

export interface Bundle<
  TValues extends { [K in keyof TOptions]: any },
  TOptions extends { [key: string]: TemplateOption }
> extends Omit<Template<TValues, TOptions>, "generate"> {
  generate(value: TValues, context: GenerateContext): Chunk[];
}
