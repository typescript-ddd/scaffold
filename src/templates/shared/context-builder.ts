import { GenerateContext, ContextImportDeclaration } from "./generate-context";
import { arrays } from "../../utils";
import { SourceFile } from "ts-morph";

export class ContextBuilder implements GenerateContext {
  private _fileName: string;
  private _importDeclarations: ContextImportDeclaration[] = [];

  private constructor(
    private rootDir: string = "@/src",
    fileName: string = "example.ts"
  ) {
    this._fileName = fileName;
  }

  resolveDir(
    type: "domain" | "infrastructure" | "application" | "presentation",
    subPath?: string
  ): string {
    switch (type) {
      case "domain":
        return `${this.rootDir}/domain${subPath ? `/${subPath}` : ""}`;
      case "infrastructure":
        return `${this.rootDir}/infrastructure${subPath ? `/${subPath}` : ""}`;
      case "application":
        return `${this.rootDir}/application${subPath ? `/${subPath}` : ""}`;
      case "presentation":
        return `${this.rootDir}/presentation${subPath ? `/${subPath}` : ""}`;
      default:
        throw new Error(`Unknown type ${type} for resolveDir`);
    }
  }

  getImportDeclarations(moduleSpecifier?: string): ContextImportDeclaration[] {
    const modules = this.getModuleSpecifiers();

    const declarations: ContextImportDeclaration[] = [];

    for (const currentModule of modules) {
      const namedImports = this.getNamedImports(currentModule);
      declarations.push({ moduleSpecifier: currentModule, namedImports });
    }

    if (moduleSpecifier) {
      return declarations.filter((x) => x.moduleSpecifier === moduleSpecifier);
    }

    return declarations;
  }

  addImportDeclaration(declaration: ContextImportDeclaration): void {
    this._importDeclarations.push(declaration);
  }

  getModuleSpecifiers(): string[] {
    return arrays.distinct(
      this._importDeclarations.map((x) => x.moduleSpecifier)
    );
  }

  getNamedImports(moduleSpecifier: string): string[] {
    return arrays.distinct(
      this._importDeclarations
        .filter((x) => x.moduleSpecifier === moduleSpecifier)
        .map((x) => x.namedImports)
        .flat()
    );
  }

  writeSourceFileImports(sourceFile: SourceFile) {
    const importDeclarations = this.getImportDeclarations();
    const moduleSpecifiers = this.getModuleSpecifiers();

    const coreImports = importDeclarations.filter(
      (x) => x.moduleSpecifier === "@typescript-ddd/core"
    );

    if (coreImports) {
      sourceFile.addImportDeclaration({
        moduleSpecifier: `@typescript-ddd/core`,
        namedImports: coreImports.map((x) => x.namedImports).flat(),
      });
    }

    // Non core and root imports
    for (const moduleSpecifier of moduleSpecifiers
      .filter(
        (x) => x !== "@typescript-ddd/core" && !x.startsWith(this.rootDir)
      )
      .sort((a, b) => {
        if (a < b) return -1;
        if (a > b) return 1;
        return 0;
      })) {
      sourceFile.addImportDeclaration({
        moduleSpecifier,
        namedImports: this.getNamedImports(moduleSpecifier),
      });
    }

    // Root imports
    for (const moduleSpecifier of moduleSpecifiers
      .filter((x) => x.startsWith(this.rootDir))
      .sort((a, b) => {
        if (a < b) return -1;
        if (a > b) return 1;
        return 0;
      })) {
      sourceFile.addImportDeclaration({
        moduleSpecifier,
        namedImports: this.getNamedImports(moduleSpecifier),
      });
    }
  }

  clearImportDeclarations(): void {
    this._importDeclarations = [];
  }

  tokenize(text: string, tokens: { name: string }[]): string {
    let output = text;
    tokens.forEach((token) => {
      output = output.replace(`{{${token.name}}}`, "${" + token.name + "}");
    });
    return "`" + output + "`";
  }

  get fileName(): string {
    return this._fileName;
  }

  static build(rootDir: string): ContextBuilder {
    return new ContextBuilder(rootDir);
  }
}
