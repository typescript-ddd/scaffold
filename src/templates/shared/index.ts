export * from "./context-builder";
export * from "./generate-context";