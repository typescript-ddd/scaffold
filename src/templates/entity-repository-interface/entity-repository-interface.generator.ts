import { Project } from "ts-morph";
import { strings } from "../../utils";
import { GenerateContext } from "../shared";
import { EntityRepositoryInterfaceTemplateValues } from "./entity-repository-interface.types";

export const generateEntityRepositoryInterface = (
  values: EntityRepositoryInterfaceTemplateValues,
  context: GenerateContext
) => {
  const { entityName } = values;
  const entityType = strings.capitalize(entityName);
  const entityTypeId = `${entityType}Id`;
  const repositoryType = `${entityType}Repository`;

  const project = new Project();
  const sourceFile = project.createSourceFile(
    `${context.fileName}.ts`,
    `
export interface ${repositoryType} extends Repository<${entityTypeId},${entityType}> {}    
`
  );

  context.addImportDeclaration({
    moduleSpecifier: `@typescript-ddd/core`,
    namedImports: ["Repository"],
  });

  context.addImportDeclaration({
    moduleSpecifier: context.resolveDir("domain"),
    namedImports: [entityType, entityTypeId],
  });

  context.writeSourceFileImports(sourceFile);

  const interfaceDeclaration = sourceFile.getInterface(repositoryType);

  interfaceDeclaration.addJsDoc({
    description: `Represents ${strings.describe(repositoryType)} interface.`,
  });

  let output = sourceFile.getFullText();

  return output;
};
