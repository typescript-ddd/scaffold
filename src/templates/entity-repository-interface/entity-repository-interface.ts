import { Project, Scope } from "ts-morph";
import { strings } from "../../utils";
import { GenerateContext } from "../shared";

export type GenerateEntityRepositoryInterfaceOptions = {
  entityName: string;
};

export const generateEntityRepositoryInterface = (
  options: GenerateEntityRepositoryInterfaceOptions,
  context: GenerateContext
) => {
  const { entityName } = options;
  const entityType = strings.capitalize(entityName);
  const entityTypeId = `${entityType}Id`;
  const repositoryType = `${entityType}Repository`;

  const project = new Project();
  const sourceFile = project.createSourceFile(
    `${context.fileName}.ts`,
    `
export interface ${repositoryType} extends Repository<${entityTypeId},${entityType}> {}    
`
  );

  context.addImportDeclaration({
    moduleSpecifier: `@typescript-ddd/core`,
    namedImports: ["Repository"],
  });

  context.addImportDeclaration({
    moduleSpecifier: context.resolveDir("domain"),
    namedImports: [entityType, entityTypeId],
  });

  context.writeSourceFileImports(sourceFile);

  const interfaceDeclaration = sourceFile.getInterface(repositoryType);

  interfaceDeclaration.addJsDoc({
    description: `Represents ${strings.describe(repositoryType)} interface.`,
  });

  let output = sourceFile.getFullText();

  return output;
};
