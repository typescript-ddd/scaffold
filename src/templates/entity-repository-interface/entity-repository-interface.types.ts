import { TemplateStringOption } from "../shared/template";

export type EntityRepositoryInterfaceTemplateValues = {
  entityName: string;
};

export type EntityRepositoryInterfaceTemplateOptions = {
  entityName: TemplateStringOption;
};
