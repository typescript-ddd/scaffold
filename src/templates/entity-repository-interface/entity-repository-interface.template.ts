import { Chunk, GenerateContext } from "../shared";
import { Template } from "../shared/template";
import { generateEntityRepositoryInterface } from "./entity-repository-interface.generator";
import {
  EntityRepositoryInterfaceTemplateOptions,
  EntityRepositoryInterfaceTemplateValues,
} from "./entity-repository-interface.types";

export class EntityRepositoryInterfaceTemplate
  implements
    Template<EntityRepositoryInterfaceTemplateValues, EntityRepositoryInterfaceTemplateOptions>
{
  readonly name = "Entity Repository Interface";
  readonly description = "Generates an entity repository interface";
  readonly options: EntityRepositoryInterfaceTemplateOptions = {
    entityName: {
      name: "entityName",
      label: "Entity Name",
      description: "The name of the entity",
      optionType: "string",
    },
  };
  readonly defaultValues = {
    entityName: "User",
  };
  generate(values: EntityRepositoryInterfaceTemplateValues, context: GenerateContext, chunkName?: string): Chunk {
    return generateEntityRepositoryInterface(values, context, chunkName);
  }
}
