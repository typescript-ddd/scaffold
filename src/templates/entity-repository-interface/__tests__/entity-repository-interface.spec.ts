import { ContextBuilder, GenerateContext } from "../../shared";
import { generateEntityRepositoryInterface } from "../entity-repository-interface";

describe("EntityRepositoryInterface", () => {
    let context: GenerateContext;
    
    beforeEach(() => {
        context = ContextBuilder.build("@src");
    });
    
    it("Should render template", () => {
        const output = generateEntityRepositoryInterface(
        { entityName: "User" },
        context
        );
        expect(output).toMatchSnapshot();
    });
});