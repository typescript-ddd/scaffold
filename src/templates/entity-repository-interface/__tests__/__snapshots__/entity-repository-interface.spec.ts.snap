// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntityRepositoryInterface Should render template 1`] = `
"import { Repository } from "@typescript-ddd/core";
import { User, UserId } from "@src/domain";

/** Represents a user repository interface. */
export interface UserRepository extends Repository<UserId,User> {}    
"
`;
